File: scripts/parse_tlds.py
#!/usr/bin/env python3
FILE = "public_suffix_list.dat"
URL = "https://publicsuffix.org/list/" + FILE
DISABLED_TLDS = (
    "exe", "pdf", "zip",
)

import sys
from imp import reload
reload(sys)
if hasattr(sys, "setdefaultencoding"):
    sys.setdefaultencoding('utf-8')
else:
    import codecs
    sys.stdout = codecs.getwriter('utf8')(sys.stdout.detach())

import os, os.path as osp, re

FILE = list(filter(osp.exists, (
    FILE,
    osp.join("scripts", FILE),
    FILE + ".txt",
    osp.join("scripts", FILE + ".txt"),
)))
FILE = FILE and FILE[0]
if FILE:
    fp = open(FILE, "rb")
else:
    import urllib.request
    fp = urllib.request.urlopen(URL)
with fp:
    lines = [line.strip().decode("utf-8") for line in fp]
    lines = list(line
        for line in lines
        if line
            and line[0] != "#"
            and line[0:2] != "//"
    )

tlds = set(suffix.split(".")[-1] for suffix in lines)
tlds -= set(DISABLED_TLDS)

tlds = list(("%02d%s" % (len(i), i), i) for i in tlds)
tlds.sort(key=lambda i: i[0])
tlds = list(i[1] for i in tlds)

prefix, tail = '  , "', ' // char[%d][%d]'
format = '%s%s"%s'
for isEn in (True, False):
    print('Utils.%s = [""' % ('_tlds' if isEn else '_nonENTlds'))
    i, count, len_tld, line, len_line = "", 0, 2, "", len(prefix)
    for i in tlds:
        if (re.match(r'^[0-9A-Za-z]+\Z', i) is not None) != isEn:
            continue
        leni = len(i)
        if leni > len_tld:
            print(format % (prefix, line, tail % (count, len_tld) if count else ""))
            line = ''
            while len_tld + 2 < leni:
                len_tld += 1
                line += '", "'
            if line:
                print(format % (prefix, line, ""))
            count, len_tld, line, len_line = 0, leni, "", len(prefix)
        len_line += leni + 1
        if len_line > (120 if isEn else 80):
            line += "\\\n." + i
            len_line = leni + 1
        else:
            line += '.' + i
        count += 1
    if count > 0:
        print(format % (prefix, line, tail % (count, len_tld) if count else ""))
        count = 0
    print("];")
    if isEn:
        print("")

#from IPython import embed; embed()

File: scripts/parse_tlds.js
#!/usr/bin/env node
"use strict"

//@ts-check
const fs = require("fs")
const pathModule = require("path")

const FILE = "public_suffix_list.dat"
const URL = "https://publicsuffix.org/list/" + FILE
const DISABLED_TLDS = [
        "exe", "pdf", "zip",
].filter(i => i)

let curFile = [
    FILE,
    pathModule.join("scripts", FILE),
    FILE + ".txt",
    pathModule.join("scripts", FILE + ".txt"),
].filter(i => i && fs.existsSync(i))[0] || ""

/** @type { string } */
const lines = fs.existsSync(curFile)
        ? fs.readFileSync(curFile, {encoding: "utf-8"})
        : await new Promise((resolve, reject) => {
    let body = ""
    const req = (URL[4] === "s" ? require("https") : require("http")).get(URL, res => {
        res.setEncoding("utf-8")
                .on("data", chunk => body += chunk)
                .on("end", () => {
            if (res.statusCode == 200) {
                resolve(body)
            } else {
                req.off("error", reject).on("abort", () => {}).abort()
                reject(`HTTP ${res.statusCode}: ${body}`)
            }
        })
    })
    req.on("error", reject)
})
const tlds = [... new Set(lines.split(/\r\n?|\n/)
            .filter(line => line && line[0] !== "#" && line.slice(0, 2) !== "//")
            .map(line => line.split(".").slice(-1)[0])
        )]
        .filter(i => ! DISABLED_TLDS.includes(i))
        .map(i => `${i.length < 10 ? "0" : ""}${i.length}${i}`)
        .sort()
        .map(i => i.slice(2))

const prefix = '  , "'
for (const isEn of [true, false]) {
    print(`BgUtils_.${isEn ? "_tlds" : "_nonENTlds"} = [""`)
    let [count, len_tld, line, len_line] = [0, 2, "", prefix.length]
    for (const i of tlds) {
        if (/^[\dA-Za-z]+$/.test(i) != isEn) { continue }
        const leni = i.length
        if (leni > len_tld) {
            print(`${prefix}${line}"${count ? ` // char[${count}][${len_tld}]` : ""}`)
            line = ""
            while (len_tld + 2 < leni) {
                len_tld += 1
                line += '", "'
            }
            if (line) {
                print(`${prefix}${line}`)
            }
            [count, len_tld, line, len_line] = [0, leni, "", prefix.length]
        }
        len_line += leni + 1
        if (len_line > (isEn ? 120 : 80)) {
            line += "\\\n." + i
            len_line = leni + 1
        } else {
            line += '.' + i
        }
        count += 1
    }
    if (count > 0) {
        print(`${prefix}${line}${count ? ` // char[${count}][${len_tld}]` : ""}`)
        count = 0
    }
    print("];")
    if (isEn) {
        print("")
    }
}

function print() {
    console.log(...arguments)
}